File: Recusrion/c++/mergesort.cpp
#include<bits/stdc++.h>
using namespace std;
void merge(int arr[] , int low ,int mid , int high)
{
    vector<int>temp;
    int left = low;
    int right = mid+1;
    while (left<=mid && right <=high)
    {
        if(arr[left]<=arr[right])
        {
            temp.push_back(arr[left]);
            left++;
        }
        else
        {
            temp.push_back(arr[right]);
            right++;
        }

    }
    while(left<=mid)
    {
        temp.push_back(arr[left]);
        left++;
    }
    while(right<=high)
    {
        temp.push_back(arr[right]);
        right++;
    }

    for(int i=low;i<=high;i++)
    {
        arr[i] = temp[i-low];
    }

}
void merge_sort(int arr[] ,int low , int high)
{
    if(low>=high)
    {
        return;
    }
    int mid = (low+high)/2;
    merge_sort(arr, low, mid);
    merge_sort(arr, mid+1 , high);
    merge(arr , low , mid, high);
}
int main(){
    int n;
    cin>>n;
    int arr[n];
    for (int i=0;i<n;i++) {
        cin >> arr[i];
    }
    merge_sort(arr, 0, n - 1);

    for (auto item : arr) {
        cout << item << " ";
    }

}

File: Recusrion/python/mergesort.py
def merge(arr, left, mid, right):
    temp = []
    low = left
    high = mid + 1

    while low <= mid and high <= right:
        if arr[low] <= arr[high]:
            temp.append(arr[low])
            low += 1
        else:
            temp.append(arr[high])
            high += 1

    while low <= mid:
        temp.append(arr[low])
        low += 1

    while high <= right:
        temp.append(arr[high])
        high += 1

    i = left
    while i <= right:
        arr[i] = temp[i - left]
        i += 1


def merge_sort(arr, left, right):
    if left >= right:
        return

    mid = (left + right) // 2
    merge_sort(arr, left, mid)
    merge_sort(arr, mid + 1, right)
    merge(arr, left, mid, right)


if __name__ == "__main__":

    arr = [2, 1, 4, 5, 3, 1, 9, 2]
    n = len(arr)
    merge_sort(arr, 0, n - 1)

    for i in arr:
        print(i, end=" ")
